diff --git a/CMakeLists.txt b/CMakeLists.txt
index 54e6a9f..3bac35b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 3.12)
 
 project(picotool)
 
+find_package(MbedTLS REQUIRED)
+
 if (NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE Release)
 endif()
@@ -102,7 +104,6 @@ add_embedded_data_project(enc_bootloader
 
 set(ENC_BOOTLOADER_ELF ${CMAKE_BINARY_DIR}/enc_bootloader/enc_bootloader.elf)
 
-if (TARGET mbedtls)
     add_embedded_data_project(enc_bootloader_mbedtls
             PREFIX enc_bootloader_mbedtls
             SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR}/enc_bootloader
@@ -116,7 +117,6 @@ if (TARGET mbedtls)
             )
 
     set(ENC_BOOTLOADER_MBEDTLS_ELF ${CMAKE_BINARY_DIR}/enc_bootloader_mbedtls/enc_bootloader.elf)
-endif()
 
 if (NOT PICOTOOL_NO_LIBUSB)
     # compile xip_ram_perms.elf
@@ -207,7 +207,7 @@ if (NOT PICOTOOL_NO_LIBUSB)
     endif()
 endif()
 
-if (TARGET mbedtls)
+if (TARGET MbedTLS::mbedtls)
     add_custom_target(embedded_data_no_libusb DEPENDS
             ${CMAKE_CURRENT_BINARY_DIR}/enc_bootloader_elf.h
             ${CMAKE_CURRENT_BINARY_DIR}/enc_bootloader_mbedtls_elf.h)
@@ -319,9 +319,11 @@ target_link_libraries(picotool
         elf2uf2
         errors
         nlohmann_json
-        whereami)
+        whereami
+        MbedTLS::mbedtls
+        MbedTLS::mbedcrypto)
 
-if (NOT TARGET mbedtls)
+if (NOT TARGET MbedTLS::mbedtls)
     message("mbedtls not found - no signing/hashing support will be built")
     target_compile_definitions(picotool PRIVATE HAS_MBEDTLS=0)
 else()
@@ -342,7 +344,8 @@ else()
     target_compile_definitions(picotool PRIVATE HAS_LIBUSB=1)
     target_link_libraries(picotool 
         picoboot_connection_cxx
-        ${LIBUSB_LIBRARIES})
+        ${LIBUSB_LIBRARIES}
+        libusb::libusb)
 endif()
 
 if (GENERATE_FIXED_DOCS_WIDTH)
@@ -390,7 +393,7 @@ install(FILES
     DESTINATION ${INSTALL_DATADIR}
 )
 
-if (TARGET mbedtls)
+if (TARGET MbedTLS::mbedtls)
     #Install enc_bootloader_mbedtls.elf
     install(FILES
         ${ENC_BOOTLOADER_MBEDTLS_ELF}
diff --git a/bintool/CMakeLists.txt b/bintool/CMakeLists.txt
index 4e3edb5..a8717e7 100644
--- a/bintool/CMakeLists.txt
+++ b/bintool/CMakeLists.txt
@@ -1,4 +1,4 @@
-if (NOT TARGET mbedtls)
+if (NOT TARGET MbedTLS::mbedtls)
     message("lib/mbedtls submodule needs to be initialized for bintool hashing/signing")
     add_library(bintool STATIC
             bintool.cpp)
@@ -25,7 +25,7 @@ else()
     target_include_directories(bintool PUBLIC ${CMAKE_CURRENT_LIST_DIR})
 
     target_link_libraries(bintool PUBLIC
-            mbedtls
+            MbedTLS::mbedtls
             elf
             errors
             boot_picobin_headers)
diff --git a/model/model.h b/model/model.h
index f942ec8..de2b8f8 100644
--- a/model/model.h
+++ b/model/model.h
@@ -4,6 +4,13 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 #pragma once
+#ifdef __cplusplus
+#include <algorithm>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#endif
 #include "boot/uf2.h"
 #include "boot/picoboot.h"
 
@@ -39,11 +46,6 @@ typedef enum {
 
 #ifdef __cplusplus
 
-#include <algorithm>
-#include <memory>
-#include <set>
-#include <string>
-#include <utility>
 #include "addresses.h"
 #include "errors.h"
 
@@ -409,4 +411,4 @@ static bool contains_unreadable_rom(uint32_t addr, uint32_t size, const model_t&
             (addr < model->unreadable_rom_start() && addr + size > model->unreadable_rom_end());
 }
 
-#endif
\ No newline at end of file
+#endif
